File: src/org/oopsc/expression/UnaryExpression.java
package org.oopsc.expression;

import org.oopsc.*;

/**
 * Die Klasse repräsentiert einen Ausdruck mit einem unären Operator im Syntaxbaum.
 */
public class UnaryExpression extends Expression {
	public enum Operator {
		MINUS, NOT
	}

	/** Der Operator. */
	public Operator operator;

	/** Der Operand, auf den der Operator angewendet wird. */
	public Expression operand;

	/**
	 * Konstruktor.
	 *
	 * @param operator
	 *        Der Operator.
	 * @param operand
	 *        Der Operand, auf den der Operator angewendet wird.
	 * @param position
	 *        Die Position, an der dieser Ausdruck im Quelltext beginnt.
	 */
	public UnaryExpression(Operator operator, Expression operand,
			Position position) {
		super(position);
		this.operator = operator;
		this.operand = operand;
	}

	@Override
	public void refPass(SemanticAnalysis sem) throws CompileException {
		this.operand.refPass(sem);

		switch (this.operator) {
			case NOT:
				this.operand.type.check(sem, Types.boolType(),
						this.operand.position);
				break;

			case MINUS:
				this.operand.type.check(sem, Types.intType(),
						this.operand.position);
				break;

			default:
				assert false;
		}

		this.type = this.operand.type;
	}

	@Override
	public void print(TreeStream tree) {
		tree.println(this.operator
				+ (this.type == null ? "" : " : " + this.type.name()));
		tree.indent();
		this.operand.print(tree);
		tree.unindent();
	}

	@Override
	public void generateCode(CodeStream code) {
		this.operand.generateCode(code, false);
		code.println("; " + this.operator);
		code.println("MRM R5, (R2)");
		switch (this.operator) {
			case NOT:
				/* TODO Optimise. */
				code.println("MRI R6, 1");
				code.println("SUB R6, R5");
				code.println("MMR (R2), R6");
			case MINUS:
				code.println("MRI R6, 0");
				code.println("SUB R6, R5");
				code.println("MMR (R2), R6");
				break;
			default:
				assert false;
		}
	}
}

File: src/org/oopsc/expression/UnaryExpression.scala
package org.oopsc.expression

import org.oopsc._
import org.oopsc.symbol.ClassSymbol

object UnaryExpression extends Enumeration {
  type Operator = Value
  val MINUS, NOT = Value
}

/**
 * Represents an expression with an unary operator in the syntax tree.
 */
case class UnaryExpression(var operator: UnaryExpression.Operator, var operand: Expression, _position: Position) extends Expression(_position) {
  import UnaryExpression._

  override def refPass(sem: SemanticAnalysis) {
    this.operand.refPass(sem)

    this.operator match {
      case NOT =>
        this.operand.resolvedType().check(Types.boolType, this.operand.position)

      case MINUS =>
        this.operand.resolvedType().check(Types.intType, this.operand.position)
    }
  }

  override def optimPass() : Expression = {
    this.operand.optimPass() match {
      case o: BooleanLiteralExpression =>
        this.operator match {
          case NOT =>
            val value = !o.value
            BooleanLiteralExpression(value, this.position)
        }

      case o: IntegerLiteralExpression =>
        this.operator match {
          case MINUS =>
            val value = -o.value
            IntegerLiteralExpression(value, this.position)
        }

      case o: UnaryExpression =>
        if (o.operator == this.operator == MINUS) {
          /* -(-x) → x */
          o.operand
        } else if (o.operator == this.operator == NOT) {
          /* NOT (NOT x) → x */
          o.operand
        } else {
          o
        }

      case o => o
    }
  }

  override def resolvedType() : ClassSymbol = this.operand.resolvedType()

  def print(tree: TreeStream) {
    tree.println(this.operator.toString + " : " + this.resolvedType().name)
    tree.indent
    this.operand.print(tree)
    tree.unindent
  }

  def generateCode(code: CodeStream) {
    this.operand.generateCode(code, false)

    code.println("; " + this.operator)
    code.println("MRM R5, (R2)")

    this.operator match {
      case NOT =>
        code.println("XOR R5, R1")
        code.println("MMR (R2), R5")

      case MINUS =>
        code.println("MRI R6, 0")
        code.println("SUB R6, R5")
        code.println("MMR (R2), R6")
    }
  }
}